### PR TITLE
Issue #7598: Add code examples for JavadocMethdod check

### LINKED ISSUE #7598: Update doc for JavadocMethod
https://checkstyle.org/config_javadoc.html#JavadocMethod

Add example for each existing config to show what code is ok and what code cause violation.
Good example is https://checkstyle.org/config_naming.html#AbstractClassName
Look at referenced commits of #5832, to see what and how should be changed in code.

Please play with module/check by means of https://checkstyle.org/cmdline.html to understand how it works.

------------

PLEASE READ https://github.com/checkstyle/checkstyle/wiki/Good-practice-on-how-to-update-examples

### PR DESCRIPTION
**Pull request by aritrasinha108**
Issue: #7598 

This PR adds a couple of code examples to the documentation of the JavadocMethod check. 

**Example 1: With default properties of JavaDocMethod**

Web Page: 
image

CLI:
```shell
aritra@aritra-VivoBook-ASUSLaptop-X412FA-X412FA:~/Desktop/open_source/checkstyle_test$ cat config1.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC           "-//Checkstyle//DTDCheckstyleConfiguration1.3//EN"           "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <module name="TreeWalker">
        <module name="JavadocMethod"/>
    </module>
</module>
aritra@aritra-VivoBook-ASUSLaptop-X412FA-X412FA:~/Desktop/open_source/checkstyle_test$ cat Test1.java 
public class Test1 {

    /**
     * A sample function to add two numbers
     * 
     * @param number1 The first number
     * @param number2 The second number
     * @return An integer
     */
    public int addTwoNumbers(int number1, int number2) { // ok, param tags present
        return number1 + number2; // ok, return tag present
    }

    /**
     * A sample function to subtract two numbers
     * 
     * @param number1 The first number
     * @param number2 The second number
     */
    public int subtractTwoNumbers(int number1, int number2) {
        return number1 - number2; // violation, return tag missing
    }

    /**
     * A sample function to print the product two numbers
     * 
     * @param number1 The first number
     */
    public void multiplyTwoNumbers(int number1, int number2) { // violation, param tag missing
        System.out.println(number1 * number2);
    } // ok, no return tag required for a void function
aritra@aritra-VivoBook-ASUSLaptop-X412FA-X412FA:~/Desktop/open_source/checkstyle_test$ java -jar checkstyle-10.8.0-all.jar -c ./config1.xml Test1.java
Starting audit...
[ERROR] /home/aritra/Desktop/open_source/checkstyle_test/Test1.java:20: @return tag should be present and have description. [JavadocMethod]
[ERROR] /home/aritra/Desktop/open_source/checkstyle_test/Test1.java:29:53: Expected @param tag for 'number2'. [JavadocMethod]
Audit done.
Checkstyle ends with 2 errors.

```

**Example 2: Checking for only Public method and ignoring any missing param tags**

Web Page:
image

CLI:
```shell
aritra@aritra-VivoBook-ASUSLaptop-X412FA-X412FA:~/Desktop/open_source/checkstyle_test$ cat config2.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC           "-//Checkstyle//DTDCheckstyleConfiguration1.3//EN"           "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <module name="TreeWalker">
        <module name="JavadocMethod">
            <property name="accessModifiers" value="public"/>
            <property name="allowMissingParamTags" value="true"/>
        </module>
    </module>
</module>
aritra@aritra-VivoBook-ASUSLaptop-X412FA-X412FA:~/Desktop/open_source/checkstyle_test$ cat Test2.java 
public class Test2 {
    /**
     * A sample function to add two numbers
     * 
     * @param number1 The first number
     */
    private int addTwoNumbers(int number1, int number2) { // ok, param tags not necessary
        return number1 + number2; // ok, only methods with public modifier will be checked
    }

    /**
     * A sample function to subtract two numbers
     */
    public int subtractTwoNumbers(int number1, int number2) {
        return number1 - number2; // violation, return tag missing
    }
}
aritra@aritra-VivoBook-ASUSLaptop-X412FA-X412FA:~/Desktop/open_source/checkstyle_test$ java -jar checkstyle-10.8.0-all.jar -c ./config2.xml Test2.java
Starting audit...
[ERROR] /home/aritra/Desktop/open_source/checkstyle_test/Test2.java:14: @return tag should be present and have description. [JavadocMethod]
Audit done.
Checkstyle ends with 1 errors.
```

**Example 3: Checking for only private and package modifiers only**

Web Page:
image

CLI:

```shell
aritra@aritra-VivoBook-ASUSLaptop-X412FA-X412FA:~/Desktop/open_source/checkstyle_test$ cat config3.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC           "-//Checkstyle//DTDCheckstyleConfiguration1.3//EN"           "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <module name="TreeWalker">
        <module name="JavadocMethod">
            <property name="accessModifiers" value="private, package"/>
        </module>
    </module>
</module>
aritra@aritra-VivoBook-ASUSLaptop-X412FA-X412FA:~/Desktop/open_source/checkstyle_test$ cat Test3.java 
class Test3 {

    /**
     * A sample function to add two numbers
     * 
     * @param number1 The first number
     * @param number2 The second number
     */
    public int addTwoNumbers(int number1, int number2) {
        return number1 + number2; // ok, no checks made for public methods
    }

    /**
     * A sample function to subtract two numbers
     * 
     * @param number1 The first number
     * @param number2 The second number
     */
    private int subtractTwoNumbers(int number1, int number2) {
        return number1 - number2; // violation, return tag missing
    }

    /**
     * A sample function to print the product two numbers
     * 
     * @param number1 The first number
     */
    void multiplyTwoNumbers(int number1, int number2) { // violation, param tag missing
        System.out.println(number1 * number2);
    } // ok, no return tag required. Method is checked since it is package-private by
      // default
}aritra@aritra-VivoBook-ASUSLaptop-X412FA-X412FA:~/Desktop/open_source/checkstyle_test$ java -jar checkstyle-10.8.0-all.jar -c ./config3.xml Test3.java
Starting audit...
[ERROR] /home/aritra/Desktop/open_source/checkstyle_test/Test3.java:19: @return tag should be present and have description. [JavadocMethod]
[ERROR] /home/aritra/Desktop/open_source/checkstyle_test/Test3.java:28:46: Expected @param tag for 'number2'. [JavadocMethod]
Audit done.
Checkstyle ends with 2 errors.
``` 
**Example 4: Checking while ignoring any missing tags**

CLI:
```shell
aritra@aritra-VivoBook-ASUSLaptop-X412FA-X412FA:~/Desktop/open_source/checkstyle_test$ cat config6.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC           "-//Checkstyle//DTDCheckstyleConfiguration1.3//EN"           "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <module name="TreeWalker">
        <module name="JavadocMethod">
            <property name="allowMissingReturnTag" value="true"/>
        </module>
    </module>
</module>
aritra@aritra-VivoBook-ASUSLaptop-X412FA-X412FA:~/Desktop/open_source/checkstyle_test$ cat Test6.java 
public class Test6 {
    /**
     * @param number1 The first number
     */
    public int addTwoNumbers(int number1, int number2) { // violation, param tag missing
        return number1 + number2; // ok, return tag not required
    }

}
aritra@aritra-VivoBook-ASUSLaptop-X412FA-X412FA:~/Desktop/open_source/checkstyle_test$ java -jar checkstyle-10.8.0-all.jar -c ./config6.xml Test6.java
Starting audit...
[ERROR] /home/aritra/Desktop/open_source/checkstyle_test/Test6.java:5:47: Expected @param tag for 'number2'. [JavadocMethod]
Audit done.
Checkstyle ends with 1 errors.
```

**Example 5: Checking while ignoring methods which have a `Deprecated` annotation associated with them**

CLI:
```shell
aritra@aritra-VivoBook-ASUSLaptop-X412FA-X412FA:~/Desktop/open_source/checkstyle_test$ cat config4.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC           "-//Checkstyle//DTDCheckstyleConfiguration1.3//EN"           "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <module name="TreeWalker">
        <module name="JavadocMethod">
            <property name="allowedAnnotations" value="Deprecated"/>
        </module>
    </module>
</module>
aritra@aritra-VivoBook-ASUSLaptop-X412FA-X412FA:~/Desktop/open_source/checkstyle_test$ cat Test4.java 
public class Test4 {
    /**
     * @param number1 The first number
     * @return An integer
     */
    public int addTwoNumbers(int number1, int number2) { // violation, param tag missing
        return number1 + number2; // ok, return tag present
    }

    @Deprecated
    /**
     * @param number1
     */
    public void multiplyTwoNumbers(int number1, int number2) { // ok, @Deprecated annotated
        System.out.println(number1 * number2); // methods won't be checked
    }

}
aritra@aritra-VivoBook-ASUSLaptop-X412FA-X412FA:~/Desktop/open_source/checkstyle_test$ java -jar checkstyle-10.8.0-all.jar -c ./config4.xml Test4.java
Starting audit...
[ERROR] /home/aritra/Desktop/open_source/checkstyle_test/Test4.java:6:47: Expected @param tag for 'number2'. [JavadocMethod]
Audit done.
Checkstyle ends with 1 errors.
```

**Example 6: Checking only for tokens which are constructor definitions**

CLI:
```shell
aritra@aritra-VivoBook-ASUSLaptop-X412FA-X412FA:~/Desktop/open_source/checkstyle_test$ cat config5.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC           "-//Checkstyle//DTDCheckstyleConfiguration1.3//EN"           "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <module name="TreeWalker">
        <module name="JavadocMethod">
            <property name="tokens" value="CTOR_DEF"/>
        </module>
    </module>
</module>
aritra@aritra-VivoBook-ASUSLaptop-X412FA-X412FA:~/Desktop/open_source/checkstyle_test$ cat Test.java 
public class Test {
    public int a;
    public int b;

    /**
     * @param x the first number
     */
    Test(int x, int y) { // violation, param tag missing
        a = x;
        b = y;
    }

    /**
     * @param number1 The first number
     * @return An integer
     */
    public int addTwoNumbers(int number1, int number2) { // ok, only constructors checked
        return number1 + number2;
    }
}
aritra@aritra-VivoBook-ASUSLaptop-X412FA-X412FA:~/Desktop/open_source/checkstyle_test$ java -jar checkstyle-10.8.0-all.jar -c ./config5.xml Test.java
Starting audit...
[ERROR] /home/aritra/Desktop/open_source/checkstyle_test/Test.java:8:21: Expected @param tag for 'y'. [JavadocMethod]
Audit done.
Checkstyle ends with 1 errors.
```
Closes #7598

### CODE REVIEW

**Comment by Vyom-Yadav** [comment]
@aritrasinha108 Thanks a lot for the PR!
A test is failing in the CI:
> Proper commit message should adhere to the following rules:
    1) Must match one of the following patterns:
        ^Issue #\d+: .*$
        ^Pull #\d+: .*$
        ^(minor|config|infra|doc|spelling|dependency|supplemental): .*$
    2) It contains only one line of text
    3) Must not end with a period, space, or tab
    4) Commit message should be less than or equal to 200 characters
The rule broken was: 1

`Issue#7598: Add code examples for JavadocMethdod check` -> `Issue #7598: Add code examples for JavadocMethdod check`

**Comment by Vyom-Yadav** [comment]
Github, generate site

**Comment by Vyom-Yadav** [comment]
Please add the proofs of examples you have posted using the CLI, see https://github.com/checkstyle/checkstyle/pull/12775#issue-1600030015 for reference.

Also if this PR completely closes the issue, you can use Closes #WXYZ in the PR description for Github to automatically close this issue when the PR is merged.

**Comment by aritrasinha108** [comment]
@Vyom-Yadav, I had to make a few minor changes in the examples to align with the CLI results. I have pushed the changes to the PR already. Need to update the screenshots in the PR description once the site is generated for the new changes,

**Comment by Vyom-Yadav** [comment]
Github, generate site

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java [hunk 1]
```diff
@@ -150,6 +150,35 @@
...
  * &lt;module name="JavadocMethod"/&gt;
  * </pre>
+ * <p>Example:</p>
+ * <pre>
+ * &#47;**
+ *  * A sample function to add two numbers
+ *  * @param number1 The first number
+ *  * @param number2 The second number
+ *  * @return An integer
+ *  *&#47;
```
Comment (Vyom-Yadav) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java:160]: Please remove the summary i.e. `A sample function to add two numbers` and just keep the tags as CI is failing due to it. (For All)

```
...
[ERROR] [checkstyle] [ERROR] /home/vsts/work/1/s/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java:239: The Javadoc block tag '@param' should be placed at the beginning of the line. [JavadocBlockTagLocation]
...
```

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java [hunk 2]
```diff
@@ -150,6 +150,35 @@
...
+ * public int addTwoNumbers(int number1, int number2){   // ok, param tags present
+ *      return number1 + number2;    // ok, return tag present
+ * }
```
Comment (Vyom-Yadav) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java:163]: Try to keep `// ok` and `// violation` comments parallel.
```
public int addTwoNumbers(int number1, int number2){   // ok, param tags present
return number1 + number2;                                                   // ok, return tag present
}
```
If you face line length error, modify variable names, modify the example to tackle it.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java [hunk 3]
```diff
@@ -150,6 +150,35 @@
  * <pre>
  * &lt;module name="JavadocMethod"/&gt;
  * </pre>
```
Comment (Vyom-Yadav) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java:152]: Please add examples for all the properties of the check. https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/108a89b_2023172613/config_javadoc.html#JavadocMethod_Properties
Some are there, add for the rest.

**Review by Vyom-Yadav — CHANGES_REQUESTED** [review]
It would be awesome if you could use the same code examples across multiple configs (even existing ones).

This would allow users to see how changes in configuration change violations, etc. Would result in better reading and understandibility.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java (hunk 3 above)
Comment (aritrasinha108) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java:152]: @Vyom-Yadav I have added the examples for the remaining properties and added their CLI test results to the PR description.

**Comment by romani** [comment]
Keep single commit always in PR.

**Comment by aritrasinha108** [comment]
> Keep single commit always in PR.
> 
> Do `mvn verify ` before push. Fix all error

@romani  I squashed the commits into one and ran ```mvn clean verify``` on my local system.

**Comment by aritrasinha108** [comment]
> Keep single commit always in PR.
> 
> Do `mvn verify ` before push. Fix all error

@romani  I squashed the commits into one and ran ```mvn clean verify``` on my local system.

**Comment by aritrasinha108** [comment]
> Keep looking at CI failures. We merge only on green CI state.
> 
> Update archunit file , as suggested at https://dev.azure.com/romanivanovjr/romanivanovjr/_build/results?buildId=13405&view=logs&j=c902ebb4-c9f8-5f09-4e17-ff78fbbc842e&t=9ca98c81-ff64-58f0-9d03-a23ac1c4a111&l=1237

@romani I updated the archunit file and most of the CI checks have passed barring 3. I could not understand what exactly was the problem with them. Could you help me fix it?

**Comment by romani** [comment]
There are more failures, please keep fixing them.

**Comment by romani** [comment]
GitHub, generate website

File: src/xdocs/config_javadoc.xml [hunk 4]
```diff
@@ -782,7 +807,22 @@ public int checkReturnTag(final int aTagIndex,
   &lt;property name=&quot;allowMissingParamTags&quot; value=&quot;true&quot;/&gt;
 &lt;/module&gt;
         </source>
+        <p>Example: </p>
+        <source>
+/**
+ * @param param1 The first number
+ */
+private int addTwoNumbers(int param1, int param2){  // ok, param tags not necessary
```
Comment (romani) [inline src/xdocs/config_javadoc.xml:815]: Please keep java code identical except for trailing comments that mark violation or explain why ok.
Java code is same, Config is changing , trailing comments are changing.

If you have problems with line length, please keep name of method and parameters as foo, p1, p2,....
It is better to put training comment on same line that actual checkstyle is placing (see CLI out put)

**Review by romani — CHANGES_REQUESTED** [review]
Items

**Comment by romani** [comment]
GitHub, generate website

**Comment by aritrasinha108** [comment]
> @aritrasinha108, please rebase one more time in latest master, to let to trigger website creation

@romani Done. I have rebased again

**Comment by aritrasinha108** [comment]
> One more time... Java code should be identical in all examples blocks. Changes are only in config and trailing comments in java code Look at https://checkstyle.org/config_design.html#MutableException for good example.

Alright I'll look into it right away

**Comment by Vyom-Yadav** [comment]
Github, generate site

File: src/xdocs/config_javadoc.xml [hunk 5]
```diff
@@ -792,6 +832,103 @@ public int checkReturnTag(final int aTagIndex,
...
+ */
+private int subtractTwoNumbers(int number1, int number2){
+    return number1 - number2;                    // violation, return tag missing
+}
+/**
+ * @param number1 The first number
+ */
+void printProduct(int number1, int number2){    // violation, param tag missing
+    System.out.println(number1 * number2);      // ok, no return tag required. Method is checked
+}                                               // since it is package-private by default
```
Comment (Vyom-Yadav) [inline src/xdocs/config_javadoc.xml:857]: Please add a new line between the methods. Also please add an example of violation for `package-private` method i.e. access modifier not specified explicitly.

File: src/xdocs/config_javadoc.xml [hunk 6]
```diff
@@ -792,6 +832,103 @@ public int checkReturnTag(final int aTagIndex,
...
+    return number1+number2;                                // ok, return tag present
+}
+
+@Deprecated
+/**
+ * @param number1
+ */
+public void multiplyTwoNumbers(int number1,int number2){   // ok, @Deprecated annotated
+    System.out.println(number1*number2);                   // methods won't be checked
+}
```
Comment (Vyom-Yadav) [inline src/xdocs/config_javadoc.xml:900]: The annotation should be after javadoc i.e.:
```java
/**                                                                                     
 * @param number1                                                                       
 */                                                                                     
@Deprecated                                                                             
public void multiplyTwoNumbers(int number1,int number2){   // ok, @Deprecated annotated 
    System.out.println(number1*number2);                   // methods won't be checked  
}                                                                                       
```

File: src/xdocs/config_javadoc.xml [hunk 7]
```diff
@@ -792,6 +832,103 @@ public int checkReturnTag(final int aTagIndex,
...
+    &lt;property name="tokens" value="CTOR_DEF"/&gt;
+&lt;/module&gt;
+        </source>
+        <p> Example:</p>
+        <source>
+public class Test {
+    public int a;
+    public int b;
+    /**
+     * @param x the first number
```
Comment (Vyom-Yadav) [inline src/xdocs/config_javadoc.xml:916]: New line after `public int b;`

**Review by Vyom-Yadav — CHANGES_REQUESTED** [review]
Please use one code example across all the configs. Don't worry if the code example becomes lengthy. A good example is at https://checkstyle.sourceforge.io/config_coding.html#VariableDeclarationUsageDistance_Examples

File: src/xdocs/config_javadoc.xml (hunk 5 above)
Comment (aritrasinha108) [reply src/xdocs/config_javadoc.xml]: @Vyom-Yadav I have removed the access modifier `public` from the `class Test` that I have been using for other examples to cover the `package-private` method use-case

Comment (aritrasinha108) [reply src/xdocs/config_javadoc.xml]: > Please use one code example across all the configs. Don't worry if the code example becomes lengthy. A good example is at https://checkstyle.sourceforge.io/config_coding.html#VariableDeclarationUsageDistance_Examples

@Vyom-Yadav  I have made all the code examples the same for every property (except for one use case of `package-private`). Could you please look into it. Thanks

**Comment by aritrasinha108** [comment]
> One more time... Java code should be identical in all examples blocks. Changes are only in config and trailing comments in java code Look at https://checkstyle.org/config_design.html#MutableException for good example.

@romani pushed the changes as you suggested.

**Comment by romani** [comment]
Please do indentation on examples to be +2 instead on yours +4.

Please remove fields, they are not required, constructor can be empty to ignore parameters.

Reduce amount of parameters to one in methods.

Remove code from methods, as it is not required and it is noise. We need to keep user focus on what matters.

Move all trailing comments close to code, it is hard to read examples on small screens (phone).

Make method names to be 3 character: boo, foo, bar,.... Their names does not matter at all.

We need to make code shorter horizontally as much as possible and reasonable.

**Comment by aritrasinha108** [comment]
> Please do indentation on examples to be +2 instead on yours +4.
> 
> Please remove fields, they are not required, constructor can be empty to ignore parameters.
> 
> Reduce amount of parameters to one in methods.
> 
> Remove code from methods, as it is not required and it is noise. We need to keep user focus on what matters.
> 
> Move all trailing comments close to code, it is hard to read examples on small screens (phone).
> 
> Make method names to be 3 character: boo, foo, bar,.... Their names does not matter at all.
> 
> We need to make code shorter horizontally as much as possible and reasonable.

@romani Made all the changes you suggested. Please check

**Comment by romani** [comment]
Github, generate site

**Comment by romani** [comment]
Screenshot_20230308-173618

Remove system.out.print. make it no codeine body.

Move trailing comments more to left, as close as possible.

**Comment by aritrasinha108** [comment]
> Screenshot_20230308-173618
> 
> Remove system.out.print. make it no codeine body.
> 
> Move trailing comments more to left, as close as possible.

@romani  Done

**Comment by romani** [comment]
Github, generate site

**Comment by romani** [comment]
```
/**
   *
   */
  void bar(int p1) {        // violation, param tag missing for p1
  }
```

Please fix comment formatting on this method in all examples to be same.

All other is good.

**Comment by romani** [comment]
Please send fixes for others issues while we do final polishing here.

**Comment by aritrasinha108** [comment]
> ```
> /**
>    *
>    */
>   void bar(int p1) {        // violation, param tag missing for p1
>   }
> ```
> 
> Please fix comment formatting on this method in all examples to be same.
> 
> All other is good.

@romani Done.

**Comment by romani** [comment]
Github, generate site

**Review by romani — APPROVED** [review]
Ok to merge

**Pull request merged**